### ISSUE TITLE
--resume issue...!

### ISSUE THREAD

**Issue by babbu3682** [Sara-Ahmed/SiT#4]
When I train again using "--resume", the Nan value appears in the model both SSL and fine-tuning. I checked the image entering input, but there was no problem.

The following will stop and warning. "Loss is nan, stopping training"

**Comment by babbu3682** [comment]
I think this is the same as my issue: https://github.com/pytorch/pytorch/issues/40497

**Comment by Sara-Ahmed** [comment]
Would you please share your log file from the main training? I suspect that the issue started from the main training and then you noticed it when you tried to resume. 
As suggested by the issue you referred to, you can disable the autocast but this will slow down the training process so I would suggest first to play with the hyper-parameters, and looking at the log file will hopefully give us a clue in which direction to move.

**Comment by babbu3682** [comment]
@Sara-Ahmed here is my log.txt from SSL

{"train_lr": 1.0000000000000887e-06, "train_loss": 3.9057014376237236, "train_RotationLoss": 1.01161359050066, "train_RotationScalar": 0.9959498764770953, "train_ContrastiveLoss": 5.41041875916944, "train_ContrastiveScalar": 1.0035055539605082, "train_ReconstructionLoss": 0.27503184485258725, "train_ReconstructionScalar": 0.9962476970808568, "test_loss": 0.8382949276964493, "test_acc1": 53.65551093931469, "test_acc5": 100.00000762553897, "epoch": 0, "n_parameters": 86000392}
{"train_lr": 1.0000000000000887e-06, "train_loss": 3.7731617456422404, "train_RotationLoss": 0.9221479808363099, "train_RotationScalar": 0.9883918963051413, "train_ContrastiveLoss": 5.213186493829357, "train_ContrastiveScalar": 1.0105134222948897, "train_ReconstructionLoss": 0.20598783211154428, "train_ReconstructionScalar": 0.9888006401636085, "test_loss": 0.7809235725470097, "test_acc1": 56.46269949875069, "test_acc5": 100.00000762553897, "epoch": 1, "n_parameters": 86000392}
{"train_lr": 0.00013361249999999244, "train_loss": 2.8431181690673766, "train_RotationLoss": 0.6699657624914539, "train_RotationScalar": 0.5389728930773353, "train_ContrastiveLoss": 4.741225427420767, "train_ContrastiveScalar": 1.4201437551267226, "train_ReconstructionLoss": 0.15360049491994252, "train_ReconstructionScalar": 0.49612313429715416, "test_loss": 0.43575159095951754, "test_acc1": 78.92513796168558, "test_acc5": 100.00000762553897, "epoch": 2, "n_parameters": 86000392}
{"train_lr": 0.00026622500000002107, "train_loss": 1.674826345294856, "train_RotationLoss": 0.43069261464820785, "train_RotationScalar": -0.12922636257727174, "train_ContrastiveLoss": 4.47270812947415, "train_ContrastiveScalar": 2.1087267121690063, "train_ReconstructionLoss": 0.09094031896239972, "train_ReconstructionScalar": -0.9374670742249676, "test_loss": 0.29722875950501815, "test_acc1": 85.81707883198837, "test_acc5": 100.00000762553897, "epoch": 3, "n_parameters": 86000392}
{"train_lr": 0.00039883750000002167, "train_loss": 0.9240683241695502, "train_RotationLoss": 0.34257443366999724, "train_RotationScalar": -0.373359862504695, "train_ContrastiveLoss": 4.347438208571325, "train_ContrastiveScalar": 2.1627198478637495, "train_ReconstructionLoss": 0.05140693981690506, "train_ReconstructionScalar": -2.7505527288147325, "test_loss": 0.23109148128314358, "test_acc1": 89.59738476506115, "test_acc5": 100.00000762553897, "epoch": 4, "n_parameters": 86000392}
{"train_lr": 0.0005314499999999893, "train_loss": 0.7188102201269916, "train_RotationLoss": 0.3070904216801953, "train_RotationScalar": -0.48537232125069785, "train_ContrastiveLoss": 4.294629549659072, "train_ContrastiveScalar": 2.150516006966348, "train_ReconstructionLoss": 0.039823906055527825, "train_ReconstructionScalar": -3.223519439908732, "test_loss": 0.20189183501112143, "test_acc1": 91.17199867525854, "test_acc5": 100.00000762553897, "epoch": 5, "n_parameters": 86000392}
{"train_lr": 0.0006639005446364931, "train_loss": 0.6462882588167869, "train_RotationLoss": 0.2923380521282811, "train_RotationScalar": -0.5364010527561739, "train_ContrastiveLoss": 4.271086298483697, "train_ContrastiveScalar": 2.145013036476221, "train_ReconstructionLoss": 0.036337265291369956, "train_ReconstructionScalar": -3.315475393097314, "test_loss": 0.20150571149192914, "test_acc1": 90.9855717561679, "test_acc5": 100.00000762553897, "epoch": 6, "n_parameters": 86000392}
{"train_lr": 0.000663829292682554, "train_loss": 0.5585539179129789, "train_RotationLoss": 0.2732773045750155, "train_RotationScalar": -0.6024799380492489, "train_ContrastiveLoss": 4.240724715095461, "train_ContrastiveScalar": 2.137892568991989, "train_ReconstructionLoss": 0.03283129579008412, "train_ReconstructionScalar": -3.4152551840607153, "test_loss": 0.16864279247995548, "test_acc1": 92.6914550132482, "test_acc5": 100.00000762553897, "epoch": 7, "n_parameters": 86000392}
{"train_lr": 0.0006637450924506043, "train_loss": 0.49553618864402726, "train_RotationLoss": 0.26053071883852313, "train_RotationScalar": -0.6504090488029977, "train_ContrastiveLoss": 4.222137412348989, "train_ContrastiveScalar": 2.1334928424483888, "train_ReconstructionLoss": 0.030489110365183626, "train_ReconstructionScalar": -3.4898524437773197, "test_loss": 0.15688990476047823, "test_acc1": 93.1133036394062, "test_acc5": 100.00000762553897, "epoch": 8, "n_parameters": 86000392}
{"train_lr": 0.0006636479472513472, "train_loss": 0.45103663984098785, "train_RotationLoss": 0.25230273358202554, "train_RotationScalar": -0.6833873022220489, "train_ContrastiveLoss": 4.210336659201439, "train_ContrastiveScalar": 2.130692422414309, "train_ReconstructionLoss": 0.02888140502753079, "train_ReconstructionScalar": -3.544264446260361, "test_loss": 0.15283553474392964, "test_acc1": 93.50372171543623, "test_acc5": 100.00000762553897, "epoch": 9, "n_parameters": 86000392}
{"train_lr": 0.0006635378609045403, "train_loss": 0.4169356446090385, "train_RotationLoss": 0.24634341181029984, "train_RotationScalar": -0.7067418775093376, "train_ContrastiveLoss": 4.199861991261158, "train_ContrastiveScalar": 2.1282022970417795, "train_ReconstructionLoss": 0.0276979063905511, "train_ReconstructionScalar": -3.58610735311263, "test_loss": 0.14945665663429375, "test_acc1": 93.57028073369167, "test_acc5": 100.00000762553897, "epoch": 10, "n_parameters": 86000392}
{"train_lr": 0.0006634148377387718, "train_loss": 0.38606220645947353, "train_RotationLoss": 0.24031032734060107, "train_RotationScalar": -0.7319689497884772, "train_ContrastiveLoss": 4.194537842521829, "train_ContrastiveScalar": 2.1269327979593142, "train_ReconstructionLoss": 0.026726320695444696, "train_ReconstructionScalar": -3.621884978785675, "test_loss": 0.14221849035746112, "test_acc1": 93.73482958236735, "test_acc5": 100.00000762553897, "epoch": 11, "n_parameters": 86000392}
{"train_lr": 0.0006632788825919191, "train_loss": 0.3644311034246426, "train_RotationLoss": 0.23735799646903796, "train_RotationScalar": -0.7448688864853897, "train_ContrastiveLoss": 4.18962654475857, "train_ContrastiveScalar": 2.1257553727525282, "train_ReconstructionLoss": 0.025942418600772493, "train_ReconstructionScalar": -3.6515133766848415, "test_loss": 0.1454876995773922, "test_acc1": 93.84545298851404, "test_acc5": 100.00000762553897, "epoch": 12, "n_parameters": 86000392}



the model was trained until 12 epochs and when I tried to resume this, Nan Loss appeared in 13 epochs.

**Comment by babbu3682** [comment]
@Sara-Ahmed  I wanted to continue learning by using the resume function.
I kept trying to stop loss from becoming an inf/nan, but I couldn't find a solution except to "torch.cuda.amp.autocast (enabled=False)". Could you solve it by any chance?

**Comment by Sara-Ahmed** [comment]
Looking at your log file, the training loss is going down so fast. How many training samples do you have? 
Also I wonder, how much does "torch.cuda.amp.autocast(enabled=False)" slow down your training process?

I would start with reducing the learning rate first before disabling the autocast.

**Comment by babbu3682** [comment]
@Sara-Ahmed  

I think LR is already very low around 0.0006. Are you talking about reducing it further?

I have 120M Child X-ray Dicom images as a trainset, batch size 170, 4 GPU 48GB

**Comment by babbu3682** [comment]
@Sara-Ahmed 

Strangely enough, if I train again from the beginning to the 501 epoch without using the resume function, nan/inf loss not occur (same lr).  

Nan/inf loss seems to occur only when learning from a certain point using the resume function. I'm trying to debug it, but it's not easy to resolve.

**Comment by Sara-Ahmed** [comment]
I am trying to figure out the issue, but it is not related to SiT specifically. Training with mixed precision is tricky sometimes.
Most probably you are getting this Nan after calculating the attention by multiplying the key with query

`attn = (q @ k.transpose(-2, -1)) * self.scale`

Can you please insert a break point there to find out if the Nan occurs there? If so, you can insert some conditions there to make sure that the output of attn is less than 65K (which is the maximum that half precision can accommodate) or even you can convert it to full precision. 

What I do not get is that why you are getting NaN only after resuming! It should also appear in the training!!

**Comment by babbu3682** [comment]
@Sara-Ahmed 
I trained again from scratch and Nan/inf loss appeared the same. I lowered lr by 1/10 than 1e-3 as a temporary measure, and nan/inf loss disappeared. 

However, it is still being trained and may appear after more epochs.

**Comment by Sara-Ahmed** [comment]
Any update? should we close the issue?
